File: scripts/plugin/serve.mjs
#!/usr/bin/env node
/**
 * Plugin serve runner: flock ledger then launch herdweb.
 * Lock file is never unlinked. The lock fd is inherited by the service child
 * so the OFD outlives the runner.
 */
import { spawn, spawnSync } from 'node:child_process'
import {
	closeSync,
	existsSync,
	mkdirSync,
	openSync,
	readFileSync,
	renameSync,
	writeFileSync,
} from 'node:fs'
import { join, resolve } from 'node:path'

const LOCK_NAME = 'herdweb.lock'
const OWNER_NAME = 'herdweb.owner.json'
const CONFIG_NAME = 'herdweb.config.ts'
const DEFAULT_CONFIG = 'export default {}\n'
const FLOCK_PY =
	'import fcntl, sys\ntry:\n    fcntl.flock(3, fcntl.LOCK_EX | fcntl.LOCK_NB)\nexcept BlockingIOError:\n    sys.exit(2)\n'
const FLOCK_PL =
	'use Fcntl qw(LOCK_EX LOCK_NB); open(FH, "<&=3") or exit(1); flock(FH, LOCK_EX|LOCK_NB) or exit(2);'

function fail(code, message) {
	console.error(message)
	process.exit(code)
}

function requiredDir(name) {
	const value = process.env[name]
	if (!value) fail(1, `ERROR ${name} is not set`)
	return value
}

function parsePort(raw) {
	if (raw === undefined || raw === '') return undefined
	const port = Number(raw)
	if (!Number.isInteger(port) || port < 1 || port > 65535) return undefined
	return port
}

function resolvePort() {
	const raw = process.env.HERDWEB_PLUGIN_PORT
	if (raw === undefined) return 7681
	const port = parsePort(raw)
	if (port === undefined) fail(1, `ERROR invalid HERDWEB_PLUGIN_PORT: ${raw}`)
	return port
}

function processStarttime(pid) {
	if (process.platform === 'linux') {
		try {
			const text = readFileSync(`/proc/${pid}/stat`, 'utf8')
			const rest = text.slice(text.lastIndexOf(')') + 2).split(/\s+/)
			return rest[19] ?? ''
		} catch {
			return ''
		}
	}
	const result = spawnSync('ps', ['-p', String(pid), '-o', 'lstart='], { encoding: 'utf8' })
	return result.status === 0 ? result.stdout.trim() : ''
}

function readOwner(stateDir) {
	try {
		const parsed = JSON.parse(readFileSync(join(stateDir, OWNER_NAME), 'utf8'))
		return parsed && typeof parsed === 'object' ? parsed : undefined
	} catch {
		return undefined
	}
}

function ownerIsTrusted(owner) {
	if (!owner || typeof owner.pid !== 'number') return false
	const recorded = String(owner.starttime ?? '')
	if (!recorded) return false
	if (process.platform === 'linux') {
		const current = processStarttime(owner.pid)
		return current !== '' && current === recorded
	}
	const alive = spawnSync('kill', ['-0', String(owner.pid)])
	if (alive.status !== 0) return false
	const lstart = processStarttime(owner.pid)
	return lstart !== '' && lstart === recorded
}

function reportLockHeld(stateDir) {
	const owner = readOwner(stateDir)
	if (ownerIsTrusted(owner)) {
		fail(2, `LOCK_HELD pid=${owner.pid} mode=${owner.mode} port=${owner.port}`)
	}
	fail(2, 'LOCK_HELD (owner metadata untrusted)')
}

function writeOwner(stateDir, payload) {
	const target = join(stateDir, OWNER_NAME)
	const tmp = `${target}.tmp`
	writeFileSync(tmp, `${JSON.stringify(payload, null, 2)}\n`)
	renameSync(tmp, target)
}

function flockNonblocking(fd) {
	const opts = { stdio: ['ignore', 'ignore', 'pipe', fd], encoding: 'utf8' }
	const py = spawnSync('python3', ['-c', FLOCK_PY], opts)
	if (py.status === 0) return true
	if (py.status === 2) return false
	const pl = spawnSync('perl', ['-e', FLOCK_PL], opts)
	if (pl.status === 0) return true
	if (pl.status === 2) return false
	fail(1, 'ERROR python3 or perl is required to take flock')
}

function openLockFile(lockPath) {
	try {
		return openSync(lockPath, 'a+')
	} catch (error) {
		const code = error && typeof error === 'object' && 'code' in error ? error.code : ''
		if (code === 'EACCES' && existsSync(lockPath)) {
			try {
				return openSync(lockPath, 'r')
			} catch (inner) {
				fail(1, `ERROR cannot open lock file: ${inner instanceof Error ? inner.message : inner}`)
			}
		}
		fail(1, `ERROR cannot open lock file: ${error instanceof Error ? error.message : error}`)
	}
}

function acquireLock(stateDir) {
	try {
		mkdirSync(stateDir, { recursive: true })
	} catch (error) {
		fail(1, `ERROR cannot create state dir: ${error instanceof Error ? error.message : error}`)
	}
	const lockPath = join(stateDir, LOCK_NAME)
	const fd = openLockFile(lockPath)
	const acquired = flockNonblocking(fd)
	if (acquired) return fd
	closeSync(fd)
	return undefined
}

function ensureConfig(configDir) {
	try {
		mkdirSync(configDir, { recursive: true })
	} catch (error) {
		fail(1, `ERROR cannot create config dir: ${error instanceof Error ? error.message : error}`)
	}
	const configPath = join(configDir, CONFIG_NAME)
	if (!existsSync(configPath)) writeFileSync(configPath, DEFAULT_CONFIG)
	return resolve(configPath)
}

function isAddrInUse(text) {
	return /EADDRINUSE|already in use/i.test(text)
}

function ownerPayload(pid, port, configPath) {
	return {
		pid,
		starttime: processStarttime(pid),
		mode: process.env.HERDWEB_PLUGIN_MODE === 'service' ? 'service' : 'pane',
		port,
		config_path: configPath,
		started_at: new Date().toISOString().replace(/\.\d{3}Z$/, 'Z'),
	}
}

function runHerdweb(pluginRoot, configPath, port, lockFd, stateDir) {
	const cliPath = join(pluginRoot, 'dist', 'cli.mjs')
	if (!existsSync(cliPath)) fail(1, `ERROR missing ${cliPath}`)
	const child = spawn(
		process.execPath,
		[cliPath, 'serve', '--config', configPath, '--port', String(port)],
		{ cwd: pluginRoot, stdio: ['inherit', 'inherit', 'pipe', lockFd] },
	)
	if (child.pid === undefined) fail(1, 'ERROR failed to spawn herdweb')
	writeOwner(stateDir, ownerPayload(child.pid, port, configPath))
	let errText = ''
	child.stderr?.on('data', (chunk) => {
		errText += chunk.toString()
		process.stderr.write(chunk)
	})
	const onSignal = (signal) => {
		child.kill(signal)
	}
	process.on('SIGINT', onSignal)
	process.on('SIGTERM', onSignal)
	child.on('exit', (code, signal) => {
		process.off('SIGINT', onSignal)
		process.off('SIGTERM', onSignal)
		if (isAddrInUse(errText)) {
			fail(3, `PORT_OCCUPIED port=${port} (got the lock; occupant is not this ledger)`)
		}
		process.exit(signal ? 1 : (code ?? 1))
	})
}

const stateDir = requiredDir('HERDR_PLUGIN_STATE_DIR')
const configDir = requiredDir('HERDR_PLUGIN_CONFIG_DIR')
const pluginRoot = process.env.HERDR_PLUGIN_ROOT || process.cwd()
const lockFd = acquireLock(stateDir)
if (lockFd === undefined) reportLockHeld(stateDir)
const configPath = ensureConfig(configDir)
const port = resolvePort()
writeOwner(stateDir, ownerPayload(process.pid, port, configPath))
runHerdweb(pluginRoot, configPath, port, lockFd, stateDir)
process.on('exit', () => {
	try {
		closeSync(lockFd)
	} catch {
		// process is exiting; child's inherited fd still holds the OFD if it is alive
	}
})

File: docs/sessions/plugin-design/probes/serve.py
#!/usr/bin/env python3
"""Minimal flock+owner.json ledger runner for herdweb plugin probes.

Binds 127.0.0.1 only. Port comes from HERDWEB_PROBE_PORT / --port (never 7681).
Mirrors design.md §2.8 start order: flock → read port → write owner.json → bind.
"""
from __future__ import annotations

import argparse
import fcntl
import json
import os
import signal
import socket
import sys
import time
from pathlib import Path

LOCK_NAME = "herdweb.lock"
OWNER_NAME = "herdweb.owner.json"

def proc_starttime(pid: int) -> str | None:
    path = Path(f"/proc/{pid}/stat")
    if not path.exists():
        return None
    text = path.read_text(errors="replace")
    rest = text[text.rfind(")") + 2 :].split()
    # field 22 starttime; after dropping pid+comm, index 19.
    return rest[19] if len(rest) > 19 else None

def owner_path(state_dir: Path) -> Path:
    return state_dir / OWNER_NAME

def read_owner(state_dir: Path) -> dict | None:
    p = owner_path(state_dir)
    if not p.is_file():
        return None
    try:
        return json.loads(p.read_text())
    except (OSError, json.JSONDecodeError) as exc:
        return {"_parse_error": str(exc)}

def diagnose_owner(owner: dict | None) -> dict:
    """Classify leftover owner.json. Callers must already know lock state."""
    if not owner or "_parse_error" in owner:
        return {"verdict": "missing_or_corrupt", "owner": owner}
    pid = owner.get("pid")
    recorded = str(owner.get("starttime") or "")
    if not isinstance(pid, int):
        return {"verdict": "malformed", "owner": owner}
    live = Path(f"/proc/{pid}").exists()
    current = proc_starttime(pid) if live else None
    if not live:
        return {"verdict": "stale_dead_pid", "owner": owner}
    if recorded and current and recorded != current:
        return {
            "verdict": "stale_pid_reused",
            "owner": owner,
            "current_starttime": current,
        }
    return {"verdict": "live", "owner": owner, "current_starttime": current}

def report_lock_held(state_dir: Path) -> int:
    diag = diagnose_owner(read_owner(state_dir))
    owner = diag.get("owner") or {}
    verdict = diag["verdict"]
    if verdict == "live":
        msg = (
            f"LOCK_HELD pid={owner.get('pid')} mode={owner.get('mode')} "
            f"port={owner.get('port')} (another herdweb is running)"
        )
    else:
        msg = (
            f"LOCK_HELD owner.json is {verdict}; cannot name the locker. "
            "lock is held but owner metadata is stale or missing"
        )
    print(msg, file=sys.stderr)
    print(json.dumps({"code": "LOCK_HELD", **diag}, ensure_ascii=False))
    return 2

def write_owner(state_dir: Path, payload: dict) -> None:
    p = owner_path(state_dir)
    tmp = p.with_suffix(".json.tmp")
    tmp.write_text(json.dumps(payload, indent=2, ensure_ascii=False) + "\n")
    tmp.replace(p)

def open_lock(state_dir: Path) -> tuple[object, bool]:
    state_dir.mkdir(parents=True, exist_ok=True)
    fh = open(state_dir / LOCK_NAME, "a+")
    try:
        fcntl.flock(fh.fileno(), fcntl.LOCK_EX | fcntl.LOCK_NB)
        return fh, True
    except BlockingIOError:
        return fh, False

def bind_loopback(port: int) -> socket.socket:
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    sock.bind(("127.0.0.1", port))
    sock.listen(8)
    return sock

def cmd_serve(args: argparse.Namespace) -> int:
    state_dir = Path(args.state_dir)
    port = args.port
    fh, ok = open_lock(state_dir)
    if not ok:
        fh.close()
        return report_lock_held(state_dir)
    pid = os.getpid()
    payload = {
        "pid": pid,
        "starttime": proc_starttime(pid),
        "mode": args.mode,
        "port": port,
        "config_path": args.config_path,
        "started_at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        "herdr_tab_id": os.environ.get("HERDR_TAB_ID"),
    }
    write_owner(state_dir, payload)
    try:
        sock = bind_loopback(port)
    except OSError as exc:
        print(
            f"PORT_OCCUPIED port={port} (got the lock; occupant is not this ledger): {exc}",
            file=sys.stderr,
        )
        print(json.dumps({"code": "PORT_OCCUPIED", "port": port, "error": str(exc)}))
        owner_path(state_dir).unlink(missing_ok=True)
        fh.close()
        return 3
    print(json.dumps({"code": "LISTENING", **payload}, ensure_ascii=False), flush=True)
    stop = False

    def _stop(signum, _frame):
        nonlocal stop
        stop = True

    signal.signal(signal.SIGTERM, _stop)
    signal.signal(signal.SIGINT, _stop)
    deadline = time.time() + args.hold_secs if args.hold_secs > 0 else None
    sock.settimeout(0.2)
    try:
        while not stop:
            if deadline is not None and time.time() >= deadline:
                break
            try:
                conn, _addr = sock.accept()
                conn.close()
            except TimeoutError:
                pass
    finally:
        sock.close()
        owner_path(state_dir).unlink(missing_ok=True)
        fh.close()
    return args.exit_code

def cmd_try_lock(args: argparse.Namespace) -> int:
    fh, ok = open_lock(Path(args.state_dir))
    print(json.dumps({"acquired": ok, "pid": os.getpid()}))
    if not ok:
        fh.close()
        return 2
    if args.hold_secs > 0:
        time.sleep(args.hold_secs)
    fh.close()
    return 0

def cmd_diagnose(args: argparse.Namespace) -> int:
    print(json.dumps(diagnose_owner(read_owner(Path(args.state_dir))), ensure_ascii=False, indent=2))
    return 0

def main() -> int:
    p = argparse.ArgumentParser()
    p.add_argument("--state-dir", default=os.environ.get("HERDR_PLUGIN_STATE_DIR", ""))
    p.add_argument("--port", type=int, default=int(os.environ.get("HERDWEB_PROBE_PORT", "17682")))
    p.add_argument("--mode", default="pane", choices=("pane", "service"))
    p.add_argument("--config-path", default=os.environ.get("HERDR_PLUGIN_CONFIG_DIR", ""))
    p.add_argument("--hold-secs", type=float, default=0)
    p.add_argument("--exit-code", type=int, default=0)
    sub = p.add_subparsers(dest="cmd", required=True)
    sub.add_parser("serve")
    sub.add_parser("try-lock")
    sub.add_parser("diagnose")
    args = p.parse_args()
    if not args.state_dir:
        print("state-dir required", file=sys.stderr)
        return 1
    if args.port == 7681:
        print("refusing port 7681", file=sys.stderr)
        return 1
    if args.cmd == "serve":
        return cmd_serve(args)
    if args.cmd == "try-lock":
        return cmd_try_lock(args)
    return cmd_diagnose(args)

if __name__ == "__main__":
    sys.exit(main())
